oxenmq/bt_producer.h

```diff
@@ -0,0 +1,306 @@
+
+using namespace std::literals;
+
+class bt_dict_producer;
+
+#if defined(__APPLE__) && defined(__MAC_OS_X_VERSION_MIN_REQUIRED) && __MAC_OS_X_VERSION_MIN_REQUIRED < 101500
+#define OXENMQ_APPLE_TO_CHARS_WORKAROUND
+/// Really simplistic version of std::to_chars on Apple, because Apple doesn't allow `std::to_chars`
+/// to be used if targetting anything before macOS 10.15.  The buffer must have at least 20 chars of
+/// space (for int types up to 64-bit); we return a pointer one past the last char written.
+template <typename IntType>
+char* apple_to_chars10(char* buf, IntType val) {
+    static_assert(std::is_integral_v<IntType> && sizeof(IntType) <= 64);
+    if constexpr (std::is_signed_v<IntType>) {
+        if (val < 0) {
+            buf[0] = '-';
+            return apple_to_chars10(buf+1, static_cast<std::make_unsigned_t<IntType>>(-val));
+        }
+    }
+
+    // write it to the buffer in reverse (because we don't know how many chars we'll need yet, but
+    // writing in reverse will figure that out).
+    char* pos = buf;
+    do {
+        *pos++ = '0' + static_cast<char>(val % 10);
+        val /= 10;
+    } while (val > 0);
+
+    // Reverse the digits into the right order
+    int swaps = (pos - buf) / 2;
+    for (int i = 0; i < swaps; i++)
+        std::swap(buf[i], pos[-1 - i]);
+
+    return pos;
+}
+#endif
+
+
+/// Class that allows you to build a bt-encoded list manually, without copying or allocating memory.
+/// This is essentially the reverse of bt_list_consumer: where it lets you stream-parse a buffer,
+/// this class lets you build directly into a buffer that you own.
+///
+/// Out-of-buffer-space errors throw 
+class bt_list_producer {
+    friend class bt_dict_producer;
+
+    // Our pointers to the next write position and the past-the-end pointer of the buffer.
+    using buf_span = std::pair<char*, char*>;
+    // Our data is a begin/end pointer pair for the root list, or a pointer to our parent if a
+    // sublist.
+    std::variant<buf_span, bt_list_producer*, bt_dict_producer*> data;
+    // Reference to the write buffer; this is simply a reference to the value inside `data` for the
+    // root element, and a pointer to the root's value for sublists/subdicts.
+    buf_span& buffer;
+    // True indicates we have an open child list/dict
+    bool has_child = false;
+    // The range that contains this currently serialized value; `from` equals wherever the `l` was
+    // written that started this list and `to` is one past the `e` that ends it.  Note that `to`
+    // will always be ahead of `buf_span.first` because we always write the `e`s to close open lists
+    // but these `e`s don't advance the write position (they will be overwritten if we append).
+    const char* const from;
+    const char* to;
+
+    // Sublist constructors
+    bt_list_producer(bt_list_producer* parent, std::string_view prefix = "l"sv);
+    bt_list_producer(bt_dict_producer* parent, std::string_view prefix = "l"sv);
```

Comment:
prefix should be `"d"sv`

Reply:
Not here, because this is constructing a list producer with a dict *parent*.  The `bt_dict_producer` constructors near the bottom call this with the `"d"sv` prefix.

Reply:
gotcha